lib/engine/devices/builtins/thingengine.builtin.ts

```diff
@@ -227,7 +227,25 @@ export default class MiscellaneousDevice extends Tp.BaseDevice {
     }
 
     do_alert() {
-        return { message_output : 'Time\'s up!'};
+        return { message_output : 'Alert' };
+    }
+
+    do_timer_expire(env : ExecWrapper) {
+        let now = new Date;
+        let duration = now.getTime() - env.app.date.getTime();
+
+        let hours = Math.floor(duration / (1000 * 60 * 60));
+        let hoursStr = hours ? `${hours} hours ` : '';
+        duration = duration % (1000 * 60 * 60);
+
+        let minutes = Math.floor(duration / (1000 * 60));
+        let minsStr = minutes ? `${minutes} minutes ` : '';
+        duration = duration % (1000 * 60);
+
+        let seconds = Math.floor(duration / (1000));
+        let secsStr = seconds ? `${seconds} seconds ` : '';
+
+        return { message_output : `It\'s been ${hoursStr} ${minsStr} ${secsStr}`};
```

Comment:
Extra spaces here?

Also, you'll want your message to be translatable, so you cannot use template strings. You can use `engine._()` to translate a message with gettext, and you can use `.format` with C-style %-specifiers (see the polyfill at https://github.com/stanford-oval/thingpedia-api/blob/master/lib/string_format.ts) or you can use [string-interp](https://github.com/stanford-oval/string-interp) which is a library for string formatting.

Reply:
If we return structured data from do_timer_expire() instead and format the phrase in manifest.tt.in, I am assuming this would no longer be necessary?